### PR TITLE
add-tagging-dropdown

### PR DESCRIPTION
**Pull request by jrp6**
JS script to create a dropdown for tagging students

### CODE REVIEW

File: django_colortag/static/add_tagging_dropdown.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/**
+ * Send an AJAX post request to create a new tagging
+ */
+function add_tagging(user_id, tag_slug, api_url) {
```
Comment (raphendyr) [inline django_colortag/static/add_tagging_dropdown.js]: Le'ts put this in private scope

File: django_colortag/static/add_tagging_dropdown.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    data: request_body,
+    contentType: 'application/json; charset=utf-8',
+    dataType: 'json',
+  });
+}
+
+/**
+ * Tag several users at once
+ */
+function add_taggings(user_ids, tag_slug, api_url) {
```
Comment (raphendyr) [inline django_colortag/static/add_tagging_dropdown.js]: And this too..

File: django_colortag/static/add_tagging_dropdown.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+/**
+ * Return f iff it is defined, and a do-nothing function otherwise
+ */
+function f_or_empty(f) {
+  return (typeof f !== 'undefined') ? f : function (x) {return;};
+}
+
+function get_create_tagging_dropdown_closure(api_url) {
```
Comment (raphendyr) [inline django_colortag/static/add_tagging_dropdown.js]: Configuration for `/usertags/` and other api url parts should be here too. You can use similar options system as I copied from somewhere to https://github.com/Aalto-LeTech/js-jquery-toggle/blob/master/jquery-toggle.js

**Comment by raphendyr** [comment]
In addition, whole js file could be inside anonymous function scope (if needed) that returns the "exported" public function. Thus, you can have syntax `(function($){ ... })(jQuery);` and thus be sure that $ is jQuery.

**Comment by raphendyr** [comment]
In addition, you can squash this change to less commits. Effectively, many of the commits are partial changes for one bigger feature, that isn't so big in the end.

**Comment by jrp6** [comment]
Removed leading slashes from default url suffixes because turns out it's easier to generate the base url with a trailing slash

File: django_colortag/static/add_tagging_dropdown.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    * @param {String} options.api_tags_url - suffix for tags, default /usertags/
+    * TODO: replace 'options' with object destructuring when supported by browsers.
+    */
+  return function get_create_tagging_dropdown_closure(options) {
+    const default_settings = {
+      api_taggings_url: 'taggings/',
+      api_tags_url: 'usertags/',
+    }
+    // TODO: replace with spread syntax (ES2018) when supported by browsers
+    const settings = Object.assign({}, default_settings, options);
```
Comment (raphendyr) [inline django_colortag/static/add_tagging_dropdown.js]: We could check that `api_url` really does end in `/` and if not, then add that.

**Comment by raphendyr** [comment]
I like it. Though, I'm not good JS coder, so I don't know. At least, there is lot of elements, that are better than in my scripts :D

File: django_colortag/models.py [hunk 5]
```diff
@@ -60,10 +61,25 @@ def __gt__(self, other):
...
+        return True
+
     def save(self, *args, **kwargs):
+        assert self.name
+
+        slug_candidate = self.slug or slugify(self.name)
+        for i in range(1000):
+            if self.is_valid_slug(slug_candidate):
+                break
+            else:
```
Comment (raphendyr) [inline django_colortag/models.py:78]: no need to have else here, because above break. At least for me it's cleaner that way.

**Pull request merged**